CryptoAnalysis/src/main/java/crypto/providerdetection/ProviderDetection.java

```diff
@@ -0,0 +1,513 @@
+					}
+				}	
+			}
+		}
+	 			
+		return rules;
+	}
+	
+	
+	
+	// Methods used from the `doAnalysis()` method
+	//-----------------------------------------------------------------------------------------------------------------
+	
+	/**
+	 * This method returns the type of Provider detected, since
+	 * it can be either `java.security.Provider` or `java.lang.String`
+	 * 
+	 * @param providerValue
+	 */
+	private String getProviderType(Value providerValue) {
+		String providerType = providerValue.getType().toString();
+		return providerType;
+	}
+	
+	
+	/**
+	 * This method return the provider used when Provider detected is of type `java.security.Provider`
+	 * 
+	 * @param statement
+	 *            
+	 * @param sootMethod
+	 *           
+	 * @param providerValue
+	 *            
+	 * @param icfg
+	 *            
+	 */
+	private String getProviderWhenTypeProvider(JAssignStmt statement, SootMethod sootMethod, Value providerValue, ObservableICFG<Unit, SootMethod> observableDynamicICFG) {
+		String provider = null;
+		
+		//Create a Boomerang solver.
+		Boomerang solver = new Boomerang(new DefaultBoomerangOptions(){
+			public boolean onTheFlyCallGraph() {
+				//Must be turned of if no SeedFactory is specified.
+				return false;
+			};
+		}) {
+			@Override
+			public ObservableICFG<Unit, SootMethod> icfg() {
+				return observableDynamicICFG;
+			}
+
+			@Override
+			public SeedFactory<NoWeight> getSeedFactory() {
+				return null;
+			}
+		};
+		Map<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> map = Maps.newHashMap();
+		for(Unit pred : observableDynamicICFG.getPredsOf(statement)) {
+			//Create a backward query
+			BackwardQuery query = new BackwardQuery(new Statement((Stmt) pred,sootMethod), new Val(providerValue, sootMethod));
+			//Submit query to the solver.
+			
+			BackwardBoomerangResults<NoWeight> backwardQueryResults = solver.solve(query);
+			map.putAll(backwardQueryResults.getAllocationSites());
+		}
+		
+		// The Provider can be correctly detected from this static analysis, if there is only one allocation site
+		// where the Provider variable was initialized. Otherwise, it throws an error because it is not possible
+		// to detect for sure the provider, if is given as parameter to the getInstance() method through the use of
+		// IF-ELSE, SWITCH statements or TERNARY operators
+		if(map.size() == 1) {
+			for(Entry<ForwardQuery, AbstractBoomerangResults<NoWeight>.Context> entry : map.entrySet()) {
+				ForwardQuery forwardQuery = entry.getKey();
+				
+				Val forwardQueryVal = forwardQuery.var();
+				Value value = forwardQueryVal.value();
+				Type valueType = value.getType();
+				String valueTypeString = valueType.toString();
+				
+				// In here are listed all the supported providers so far
+				if(valueTypeString.contains("BouncyCastleProvider")) {
+					provider = "BC";
+				}
+			}
+		}
+		else if (map.size() > 1) {
+			throw new RuntimeException("The provider parameter must be passed directly to the"
```

Comment:
Is an exception the best option here? We should rather log a warning and default to the standard JCA ruleset